src/Microsoft.Extensions.Configuration/IncludeConfigurationProvider.cs

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Configuration
+{
+    public class IncludeConfigurationProvider : ConfigurationProvider
+    {
+        public IncludeConfigurationProvider(IConfiguration source)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+            foreach (var section in source.GetChildren())
+            {
+                AddSection(section, section.Key);
+            }
+        }
+
+        private void AddSection(IConfigurationSection section, string keyPrefix)
+        {
+            Data.Add(keyPrefix, section.Value);
+            foreach (var child in section.GetChildren())
+            {
+                AddSection(child, keyPrefix + Constants.KeyDelimiter + child.Key);
```

Comment:
It is a bit of a shame that we have to rebuild the key like this every time with `keyPrefix + Constants.KeyDelimiter + child.Key`. I.e. child.Key does some processing every time and then we do the concatenation. I think we can avoid most of it (although I haven't tested it):

- If the IConfiguration passed to the constructor was an IConfigurationRoot, then `keyPrefix + Constants.KeyDelimiter + child.Key` can be replaced with just `child.Path`. 
- Otherwise it should be just `child.Path.Substring(source.Path.Lenght + 1)`, and `source.Path.Lenght + 1` remains the same for the whole provider instance so we can compute it only once.